### PR TITLE
bug-fix/internal-issue-#609

### PR DESCRIPTION
**Pull request by gnusi**
Big redesign of ArangoSearch recovery is intended to drastically simplify the code, improve maintainability and speedup the recovery. The approach is generic and may be reused in other cases where synchronization of storage engine and 3rd-party source on per transaction basis is required.

### CODE REVIEW

File: arangod/IResearch/IResearchRocksDBRecoveryHelper.cpp [hunk 1]
```diff
@@ -188,20 +188,48 @@ void ensureLink(arangodb::DatabaseFeature& db,
   }
 }
 
+class RecoveryTransactionContext final : public arangodb::transaction::StandaloneContext {
```
Comment (graetzer) [inline arangod/IResearch/IResearchRocksDBRecoveryHelper.cpp]: this code seems like its duplicated

File: arangod/RocksDBEngine/RocksDBTransactionState.h [hunk 2]
```diff
@@ -147,6 +147,15 @@ class RocksDBTransactionState final : public TransactionState {
   ///        Used to update the estimate after the trx committed
   void trackIndexRemove(TRI_voc_cid_t cid, TRI_idx_iid_t idxObjectId, uint64_t hash);
 
+  /// @returns tick of last operation in a transaction
+  /// @note not in recovery: the value is valid only after transaction is committed
+  /// @note in recovery: the value is valid immediately after transaction is started
+  TRI_voc_tick_t lastOperationTick() const noexcept final { return _lastOperationTick; }
```
Comment (graetzer) [inline arangod/RocksDBEngine/RocksDBTransactionState.h]: seems like this has about the same purpose as `_lastWrittenOperationTick` in mmfiles. Perhaps move this up into the base-class ?

Comment (gnusi) [reply arangod/RocksDBEngine/RocksDBTransactionState.h]: @graetzer sure, that is on my list :-)

**Comment by gnusi** [comment]
seems to be related
https://github.com/arangodb/backlog/issues/552

**Comment by gnusi** [comment]
https://github.com/arangodb/backlog/issues/609

**Comment by gnusi** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/

**Comment by gnusi** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/5265/

File: arangod/IResearch/IResearchLink.cpp [hunk 3]
```diff
@@ -521,99 +538,69 @@ arangodb::Result IResearchLink::commitUnsafe() {
...
-      auto& checkpoint = reader.meta().filename;
-      auto checkpointFile = checkpoint + std::string(IRESEARCH_CHECKPOINT_SUFFIX);
-      auto ref = irs::directory_utils::reference(*(_dataStore._directory), checkpointFile, true);
-      arangodb::velocypack::Builder builder;
+    // update reader
+    _dataStore._reader = reader;
 
-      builder.add(arangodb::velocypack::Value(checkpoint));
+    // invalidate query cache
+    aql::QueryCache::instance()->invalidate(&(_collection.vocbase()), _viewGuid);
```
Comment (graetzer) [inline arangod/IResearch/IResearchLink.cpp]: why is the query cache invalidated in here ?

Comment (graetzer) [reply arangod/IResearch/IResearchLink.cpp]: is this not beeing done in `TransactionState::clearQueryCache()` ?

File: arangod/IResearch/IResearchLink.cpp [hunk 4]
```diff
@@ -1000,211 +1007,112 @@ arangodb::Result IResearchLink::initDataStore(InitCallback const& initCallback,
...
+      _dataStore._inRecovery = true; // will be adjusted in post-recovery callback
+      _recoveryTick = _engine->recoveryTick();
+      break;
     }
 
-    auto& checkpointFile = *ref; // ref non-null ensured by above loop
-    auto in = _dataStore._directory->open( // open checkpoint file
-      checkpointFile, irs::IOAdvice::NORMAL // args, use 'NORMAL' since the file could be empty
-    );
+    case RecoveryState::IN_PROGRESS: { // link is being created during recovery
```
Comment (graetzer) [inline arangod/IResearch/IResearchLink.cpp]: question: Rocksdb does not re-create indexes during a recovery, indexes that were only partially build before a crash are just removed. Is that consistent with that comment ?

File: arangod/IResearch/IResearchLink.cpp [hunk 5]
```diff
@@ -307,13 +345,11 @@ void IResearchLink::afterTruncate() {
   SCOPED_LOCK(_asyncSelf->mutex());  // '_dataStore' can be asynchronously modified
 
   if (!*_asyncSelf) {
+    // the current link is no longer valid (checked after ReadLock aquisition)
```
Comment (jsteemann) [inline arangod/IResearch/IResearchLink.cpp]:
```suggestion
    // the current link is no longer valid (checked after ReadLock acquisition)
```

File: arangod/IResearch/IResearchLink.cpp [hunk 6]
```diff
@@ -367,17 +403,17 @@ void IResearchLink::batchInsert( // insert documents
...
-      LOG_TOPIC("7d258", WARN, arangodb::iresearch::TOPIC)
+      LOG_TOPIC("7d258", WARN, iresearch::TOPIC)
           << "failed to lock arangosearch link while inserting a batch into "
              "arangosearch link '"
           << id() << "', tid '" << state.id() << "'";
-      queue->setStatus(TRI_ERROR_ARANGO_INDEX_HANDLE_BAD);  // the current link is no longer
-                                                            // valid (checked after ReadLock
-                                                            // aquisition)
+
+      // the current link is no longer valid (checked after ReadLock aquisition)
```
Comment (jsteemann) [inline arangod/IResearch/IResearchLink.cpp]:
```suggestion
      // the current link is no longer valid (checked after ReadLock acquisition)
```

File: arangod/IResearch/IResearchLink.cpp [hunk 7]
```diff
@@ -476,32 +488,35 @@ arangodb::Result IResearchLink::cleanupUnsafe() {
...
       "caught exception while cleaning up arangosearch link '" + std::to_string(id()) +
-      "' run id '" + std::to_string(size_t(&runId)) + "'");
+      "' run id '" + std::to_string(size_t(&runId)) + "'"
+    };
   }
 
-  return arangodb::Result();
-}
+  LOG_TOPIC("7e821", TRACE, iresearch::TOPIC)
+    << "successfull cleanup of arangosearch link '" << id()
```
Comment (jsteemann) [inline arangod/IResearch/IResearchLink.cpp]:
```suggestion
    << "successful cleanup of arangosearch link '" << id()
```

File: arangod/IResearch/IResearchLink.cpp [hunk 8]
```diff
@@ -521,99 +536,78 @@ arangodb::Result IResearchLink::commitUnsafe() {
...
+    auto reader = _dataStore._reader.reopen();
 
     if (!reader) {
       // nothing more to do
-      LOG_TOPIC("37bcf", WARN, arangodb::iresearch::TOPIC)
+      LOG_TOPIC("37bcf", WARN, iresearch::TOPIC)
         << "failed to update snapshot after commit, run id '" << size_t(&runId)
         << "', reuse the existing snapshot for arangosearch link '" << id() << "'";
 
       return {};
```
Comment (jsteemann) [inline arangod/IResearch/IResearchLink.cpp]: mmh, now it only logs a warning but returns an `ok()` Result from here. Shouldn't it return an error?

File: arangod/IResearch/IResearchLink.cpp [hunk 9]
```diff
@@ -521,99 +536,78 @@ arangodb::Result IResearchLink::commitUnsafe() {
...
-      _dataStore._recovery_ref = ref; // ensure checkpoint file will not get removed
-    }
+    // invalidate query cache
+    aql::QueryCache::instance()->invalidate(&(_collection.vocbase()), _viewGuid);
 
-    _dataStore._reader = reader; // update reader
-    arangodb::aql::QueryCache::instance()->invalidate(&(_collection.vocbase()), _viewGuid);
-  } catch (arangodb::basics::Exception const& e) {
+    LOG_TOPIC("7e328", TRACE, iresearch::TOPIC)
+      << "successfull sync of arangosearch link '" << id()
```
Comment (jsteemann) [inline arangod/IResearch/IResearchLink.cpp]:
```suggestion
      << "successful sync of arangosearch link '" << id()
```

File: arangod/IResearch/IResearchLink.cpp [hunk 10]
```diff
@@ -636,54 +630,60 @@ arangodb::Result IResearchLink::commitUnsafe() {
...
+    return {
+      TRI_ERROR_INTERNAL,
+      "caught exception while executing consolidation policy '" + policy.properties().toString() +
+      "' on arangosearch link '" + std::to_string(id()) + "' run id '" + std::to_string(size_t(&runId)) + "'"
+    };
   }
 
-  return arangodb::Result();
+  LOG_TOPIC("7e828", TRACE, iresearch::TOPIC)
+    << "successfull consolidation of arangosearch link '" << id()
```
Comment (jsteemann) [inline arangod/IResearch/IResearchLink.cpp]:
```suggestion
    << "successful consolidation of arangosearch link '" << id()
```

File: arangod/IResearch/IResearchLink.cpp [hunk 11]
```diff
@@ -723,95 +724,94 @@ arangodb::Result IResearchLink::drop() {
...
+    auto* ci = ClusterInfo::instance();
 
     if (!ci) {
-      return arangodb::Result( // result
-        TRI_ERROR_INTERNAL, // code
-        std::string("failure to get storage engine while initializing arangosearch link '") + std::to_string(_id) + "'"
-      );
+      return {
+        TRI_ERROR_INTERNAL,
+        "failure to get storage engine while initializing arangosearch link '" + std::to_string(_id) + "'"
```
Comment (jsteemann) [inline arangod/IResearch/IResearchLink.cpp]: Error message is misleading, as it is actually not the storage engine that is missing but the cluster info.

File: arangod/IResearch/IResearchLink.cpp [hunk 12]
```diff
@@ -826,18 +826,18 @@ arangodb::Result IResearchLink::init(
...
```
Comment (jsteemann) [inline arangod/IResearch/IResearchLink.cpp]: same misleading error message here

File: arangod/IResearch/IResearchLink.cpp [hunk 13]
```diff
@@ -1459,61 +1397,48 @@ arangodb::Result IResearchLink::insert(
...
+      };
     }
 
     TRI_ASSERT(_dataStore); // must be valid if _asyncSelf->get() is valid
 
-    // optimization for single-document insert-only transactions
-    if (trx.isSingleOperationTransaction() // only for single-docuemnt transactions
-        && RecoveryState::DONE == _dataStore._recovery) {
-      auto ctx = _dataStore._writer->documents();
+//FIXME try to preserve optimization
```
Comment (jsteemann) [inline arangod/IResearch/IResearchLink.cpp]: what about this comment? does it mean the optimization is turned off?

File: arangod/RestServer/FlushFeature.cpp [hunk 14]
```diff
@@ -578,119 +74,22 @@ void FlushFeature::collectOptions(std::shared_ptr<ProgramOptions> options) {
...
     LOG_TOPIC("798c4", ERR, Logger::FLUSH) << "FlushFeature not running";
 
-    return nullptr;
+    return;
   }
 
   _flushSubscriptions.emplace_back(subscription);
 
-  return subscription;
+  return;
```
Comment (jsteemann) [inline arangod/RestServer/FlushFeature.cpp]: this return statement seems unnecessary now
```suggestion
```

File: arangod/MMFiles/MMFilesWalRecoverState.cpp [hunk 15]
```diff
@@ -1399,6 +1412,8 @@ int MMFilesWalRecoverState::replayLogfile(MMFilesWalLogfile* logfile, int number
       << "replaying WAL logfile '" << logfileName << "' (" << (number + 1)
       << " of " << n << ")";
 
+  usleep(5000000);
```
[the remark this answers is not in the document]
Comment (gnusi) [reply arangod/MMFiles/MMFilesWalRecoverState.cpp]: блядь

File: arangod/IResearch/IResearchLink.cpp (hunk 13 above)
Comment (gnusi) [reply arangod/IResearch/IResearchLink.cpp]: yes, just a reminder, currently we don't operation transaction tick in advance

File: arangod/IResearch/IResearchLink.cpp (hunk 12 above)
Comment (gnusi) [reply arangod/IResearch/IResearchLink.cpp]: done

File: arangod/IResearch/IResearchLink.cpp (hunk 11 above)
Comment (gnusi) [reply arangod/IResearch/IResearchLink.cpp]: done

File: arangod/IResearch/IResearchLink.cpp (hunk 10 above)
Comment (gnusi) [reply arangod/IResearch/IResearchLink.cpp]: done

File: arangod/IResearch/IResearchLink.cpp (hunk 7 above)
Comment (gnusi) [reply arangod/IResearch/IResearchLink.cpp]: done

File: arangod/IResearch/IResearchLink.cpp (hunk 6 above)
Comment (gnusi) [reply arangod/IResearch/IResearchLink.cpp]: done

File: arangod/IResearch/IResearchLink.cpp (hunk 5 above)
Comment (gnusi) [reply arangod/IResearch/IResearchLink.cpp]: done

File: 3rdParty/iresearch/core/index/segment_writer.hpp [hunk 16]
```diff
@@ -227,6 +227,9 @@ class IRESEARCH_API segment_writer: util::noncopyable {
   void reset() NOEXCEPT;
   void reset(const segment_meta& meta);
 
+  void tick(uint64_t tick) NOEXCEPT { tick_ = tick; }
```
[the remark this answers is not in the document]
Comment (gnusi) [reply 3rdParty/iresearch/core/index/segment_writer.hpp]: done

File: arangod/IResearch/IResearchLink.cpp (hunk 3 above)
Comment (gnusi) [reply arangod/IResearch/IResearchLink.cpp]: ArangoSearch has a number of dedicated maintenance jobs which can upgrade snapshot asynchronously. That's why we need to invalidate query cache here as well.

File: arangod/IResearch/IResearchLink.cpp (hunk 4 above)
Comment (gnusi) [reply arangod/IResearch/IResearchLink.cpp]: the comment is written with the assumption `IResearchLink == Index` and states that if recovery stepped on link creation marker, link will be created based on the data in restored collection(s).

**Comment by gnusi** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/5298/

File: arangod/RestServer/FlushFeature.cpp (hunk 14 above)
Comment (gnusi) [reply arangod/RestServer/FlushFeature.cpp]: done

**Comment by gnusi** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/5304/

**Pull request merged**